Mathlib/CategoryTheory/MorphismProperty/Composition.lean

```diff
@@ -198,6 +198,86 @@ instance inf {P Q : MorphismProperty C} [P.IsMultiplicative] [Q.IsMultiplicative
 
 end IsMultiplicative
 
+/-- Given a morphism property `W`, the `MultiplicativeClosure W` is the smallest
+multiplicative property greater or equal than `W`. -/
+inductive MultiplicativeClosure (W : MorphismProperty C) : MorphismProperty C
+  | of {x y : C} (f : x ⟶ y) (hf : W f) : MultiplicativeClosure W f
+  | id (x : C) : MultiplicativeClosure W (𝟙 x)
+  | comp_of {x y z : C} (f : x ⟶ y) (g : y ⟶ z) (hf : MultiplicativeClosure W f) (hg : W g) :
+    MultiplicativeClosure W (f ≫ g)
+
+/-- A variant of MultiplicativeClosure in which compositions are taken on the left rather than
+on the right. We show in `MultiplicativeClosure.eq_prime` that this is in fact equal
+to MultiplicativeClosure. -/
+inductive MultiplicativeClosure' (W : MorphismProperty C) : MorphismProperty C
+  | of {x y : C} (f : x ⟶ y) (hf : W f) : MultiplicativeClosure' W f
+  | id (x : C) : MultiplicativeClosure' W (𝟙 x)
+  | of_comp {x y z : C} (f : x ⟶ y) (g : y ⟶ z) (hf : W f) (hg : MultiplicativeClosure' W g) :
+    MultiplicativeClosure' W (f ≫ g)
+
+namespace MultiplicativeClosure
+
+variable {W : MorphismProperty C}
+
+/-- `MultiplicativeClosure W` is multiplicative. -/
+instance : IsMultiplicative (MultiplicativeClosure W) where
+  id_mem x := .id x
+  comp_mem f g hf hg := by
+    induction hg with
+    | of _ hf₀ => exact .comp_of f _ hf hf₀
+    | id _ => rwa [Category.comp_id]
+    | comp_of f' g hf' hg h_rec =>
+      rw [← Category.assoc]
+      exact .comp_of (f ≫ f') g (h_rec f hf) hg
+
+/-- The multiplicative closure is greater or equal than the original property. -/
+lemma self_le : W ≤ (MultiplicativeClosure W) := fun {_ _} _ hf ↦ .of _ hf
+
+/-- The multiplicative closure of a multiplicative property is equal to itself. -/
+lemma eq_self_of_isMultiplicative [W.IsMultiplicative] : (MultiplicativeClosure W) = W := by
+  apply le_antisymm _ self_le
+  intro _ _ _ hf
+  induction hf with
+  | of _ hf₀ => exact hf₀
+  | id x => exact W.id_mem x
+  | comp_of _ _ _ hg hf => exact W.comp_mem _ _ hf hg
+
+/-- The multiplicative closure of `W` is the smallest multiplicative property greater or equal than
+`W`. -/
+lemma le_of_isMultiplicative_le
```

Comment:
It would be nice to have a `multiplicativeClosure_le_iff` asserting `W.multiplicativeClosure ≤ W'` iff `W ≤ W'` under the `W'.IsMultiplicative` assumption.